cudax/examples/cccl-rt/legacy_interact/legculib.cpp

```diff
@@ -0,0 +1,340 @@
+    case LEGCULIB_ERROR_INVALID_ARGUMENT:
+      return "an invalid argument was passed to the API call";
+    case LEGCULIB_ERROR_KERNEL_COMPILATION:
+      return "a kernel compilation failed";
+    case LEGCULIB_ERROR_MALLOC:
+      return "failed to allocate memory from the heap";
+    case LEGCULIB_ERROR_CUDA_DRIVER:
+      return "a CUDA Driver error occurred";
+    default:
+      return "unknown error";
+  }
+}
+
+//! @brief Makes the success result.
+static inline legculib_result success()
+{
+  return {LEGCULIB_SUCCESS, {}};
+}
+
+//! @brief Makes the error result.
+template <legculib_error e>
+static inline legculib_result error()
+{
+  static_assert(e != LEGCULIB_ERROR_CUDA_DRIVER);
+  return {e, {}};
+}
+
+//! @brief Makes the error result for CUDA call.
+static inline legculib_result cuda_error(CUresult error)
+{
+  return {LEGCULIB_ERROR_CUDA_DRIVER, static_cast<cudaError_t>(error)};
+}
+
+#define CALL_CUDA(...)                  \
+  do                                    \
+  {                                     \
+    const auto _result = (__VA_ARGS__); \
+    if (_result != CUDA_SUCCESS)        \
+    {                                   \
+      return cuda_error(_result);       \
+    }                                   \
+  } while (false)
+#define CALL_NVRTC(...)                                  \
+  do                                                     \
+  {                                                      \
+    const auto _result = (__VA_ARGS__);                  \
+    if (_result != NVRTC_SUCCESS)                        \
+    {                                                    \
+      return error<LEGCULIB_ERROR_KERNEL_COMPILATION>(); \
+    }                                                    \
+  } while (false)
+
+//--------------------------------------------------------------------------------------------------------------------//
+// Initialization
+//--------------------------------------------------------------------------------------------------------------------//
+
+//! @brief Checks whether the library is initialized.
+static inline bool is_initialized()
+{
+  return context != nullptr;
+}
+
+//! @brief Initializes the library for the ordinal device.
+//! @note Public API.
+extern "C" legculib_result legculib_init(int ordinal)
+{
+  if (is_initialized())
+  {
+    return success();
+  }
+
+  // Initialize CUDA Driver.
+  CALL_CUDA(cuInit(0));
+
+  // Get the CUdevice from the ordinal.
+  CALL_CUDA(cuDeviceGet(&device, ordinal));
+
+  // Get the compute capability.
+  int cc_major{};
+  CALL_CUDA(cuDeviceGetAttribute(&cc_major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, device));
+  int cc_minor{};
+  CALL_CUDA(cuDeviceGetAttribute(&cc_minor, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR, device));
+  device_cc = 10 * cc_major + cc_minor;
+
+  // Retain the primary context.
+  CALL_CUDA(cuDevicePrimaryCtxRetain(&context, device));
+
+  return success();
+}
```

Comment:
FYI, only the three of us will do it like this 😆 Many library folks just do
```c++
int orig_dev;
CALL_CUDA( cudaGetDevice(&orig_dev) );
CALL_CUDA( cudaSetDevice(ordinal) );
CALL_CUDA( cudaFree(0) );
CALL_CUDA( cudaSetDevice(orig_dev) );
```
I think this example should be made as cudart-based.